### DISCUSSION TITLE
two inner CE: Singularity CE and Pool CE can work together or not?

### DISCUSSION THREAD

**Discussion by zhangxiaomei** [DIRACGrid/DIRAC#5938] [Q&A] [answered]
Hi all,
If I want to support multi-core jobs, I need to choose localCEType=Pool.  Does mean I can't use Singularity since it is another type of inner CE? Is it possible to use both? 

Best regards,
Xiaomei

**Comment by aldbr** [comment, marked as answer]
Looks like this is not really documented (at least I didn't find the information in the DIRAC documentation).

The `Pool` computing element (CE) is a special case, it embeds an inner computing element.
By default the `Pool` CE uses the `InProcess` CE, which simply executes the job on the machine.
The `Pool` CE also supports the `Singularity` CE. To use it, you just have to add the following option in the CE configuration: 

~InnerCESubmissionType = Singularity~
```
LocalCEType = Pool/Singularity
```

**Comment by zhangxiaomei** [comment]
Hi alddr,
Thank you for the reply!  I have just done the tests and it looks like only one kind of inner CE type (Pool) was selected if both inner type was defined even with "InnerCESubmissionType":
In the pilot log, I saw "Defining Inner CE from local configuration = Pool",  and singularity is not started. The CS configuration I used:
LocalCEType = Pool
NumberOfProcessors = 4
Tag = WholeNode
InnerCESubmissionType = Singularity
ContainerRoot = /cvmfs/juno.ihep.ac.cn/singularity/juno-centos7
Container = Singularity

**Reply by aldbr** [reply]
> In the pilot log, I saw "Defining Inner CE from local configuration = Pool"

This is expected, this line does not mean that the `SingularityCE` is not used.
I agree this is not very clear, the line should actually indicate "Defining local CE from local configuration".
In the pilot log, can you see this line: "Creating singularity container"?

Also, do you have any error that let you think that Singularity is not used?

**Reply by aldbr** [reply]
And where did you find the following option: `Container = Singularity`? It should not be needed.

**Reply by zhangxiaomei** [reply]
In the pilot log with multi-core setting, I didn't see the line: "Creating singularity container". Comparing the logs, the singularity part is missing in multi-core setting, eg:
"2022-03-14 04:03:10 UTC WorkloadManagement/JobAgent INFO: Found Job LogLevel JDL parameter with value: info
Starting inner container wrapper scripts (uses dirac-install) at Mon Mar 14 04:03:11 UTC 2022.
2022-03-14 04:03:11 UTC dirac-install [NOTICE]  Processing installation requirements
2022-03-14 04:03:11 UTC dirac-install [NOTICE]  Destination path for installation is /tmp
\+ cd /tmp
\+ unset DIRAC
\+ unset DIRACOS
\+ ./dirac-install.py -V IHEP -r v1r0" 
In previous version, it is needed and we haven't removed them, but it doesn't matter.
I just want to prove if it is used since the experiment asks to have that.

**Reply by aldbr** [reply]
Could you try to define `LocalCEType` as `Pool/Singularity` and let me know how it goes please?

**Reply by aldbr** [reply]
Which version of DIRAC are you using by the way?

**Reply by zhangxiaomei** [reply]
My DIRAC version v7r1p30.

**Reply by zhangxiaomei** [reply]
I have a try with "LocalCEType = Pool/Singularity".  In the end, the piolt use PoolCE, no singularity:
2022-03-15 03:34:59 UTC WorkloadManagement/JobAgent INFO: Defining Inner CE from local configuration = Pool/Singularity
2022-03-15 03:35:19 UTC WorkloadManagement/JobAgent INFO: Submitting JobWrapper Job9267834 to PoolCE

**Reply by aldbr** [reply]
I created a PR to resolve the issue: https://github.com/DIRACGrid/DIRAC/pull/5959.
`v7r1` is not supported anymore, so I suggest you to update your DIRAC installation to get the fix (at least `v7r2`).
